pkg/additionalimages/updater.go

```diff
@@ -0,0 +1,77 @@
+package additionalimages
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cloudogu/k8s-backup-operator/pkg/api/ecosystem"
+	v1 "github.com/cloudogu/k8s-backup-operator/pkg/api/v1"
+)
+
+type updater struct {
+	clientSet    ecosystemClientSet
+	namespace    string
+	kubectlImage string
+}
+
+func NewUpdater(clientSet ecosystem.Interface, namespace string, kubectlImage string) Updater {
+	return &updater{clientSet: clientSet, namespace: namespace, kubectlImage: kubectlImage}
+}
+
+// Update sets the newest additional images wherever they are needed.
+// E.g., the kubectl image used in the CronJob of a BackupSchedule.
+func (bsp *updater) Update(ctx context.Context) error {
+	backupScheduleClient := bsp.clientSet.EcosystemV1Alpha1().BackupSchedules(bsp.namespace)
+	scheduleList, err := backupScheduleClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list backup schedules whose images are not up to date: %w", err)
+	}
+
+	var errs []error
+	for _, backupSchedule := range scheduleList.Items {
+		if backupSchedule.Status.CurrentKubectlImage == bsp.kubectlImage {
+			// image is up-to-date, nothing to do
+			continue
+		}
+
+		err = bsp.patchCronJob(ctx, &backupSchedule)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to update additional images in cron job %s: %w", backupSchedule.CronJobName(), err))
+			continue
+		}
+
+		err = bsp.updateCurrentImage(ctx, &backupSchedule, backupScheduleClient)
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (bsp *updater) patchCronJob(ctx context.Context, schedule *v1.BackupSchedule) error {
+	cronJobClient := bsp.clientSet.BatchV1().CronJobs(bsp.namespace)
+	cronJob, err := cronJobClient.Get(ctx, schedule.CronJobName(), metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get cron job %s: %w", schedule.CronJobName(), err)
+	}
+
+	cronJob.Spec.JobTemplate.Spec.Template.Spec.Containers[0].Image = bsp.kubectlImage
+	cronJob, err = cronJobClient.Update(ctx, cronJob, metav1.UpdateOptions{})
```

Comment:
🚫 **[checkstyle]** <staticcheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>SA4006: this value of `cronJob` is never used